test/insertion.jl

```diff
@@ -0,0 +1,1253 @@
+        for cols in [:orderequal, :setequal, :intersect]
+            df = DataFrame(a=1:3)
+            with_logger(SimpleLogger(IOBuffer())) do
+                @test_throws MethodError push!(df, v, cols=cols)
+            end
+            @test push!(df, v, cols=cols, promote=true) == DataFrame(a=[1:3; "a"])
+
+            df = DataFrame(a=1:3)
+            with_logger(SimpleLogger(IOBuffer())) do
+                @test_throws MethodError pushfirst!(df, v, cols=cols)
+            end
+            @test pushfirst!(df, v, cols=cols, promote=true) == DataFrame(a=["a"; 1:3])
+
+            df = DataFrame(a=1:3)
+            with_logger(SimpleLogger(IOBuffer())) do
+                @test_throws MethodError insert!(df, 2, v, cols=cols)
+            end
+            @test insert!(df, 2, v, cols=cols, promote=true) == DataFrame(a=[1, "a", 2, 3])
+
+        end
+        for cols in [:subset, :union]
+            df = DataFrame(a=1:3, b=11:13)
+            with_logger(SimpleLogger(IOBuffer())) do
+                @test_throws MethodError push!(df, v, cols=cols, promote=false)
+            end
+            @test push!(df, v, cols=cols) ≅ DataFrame(a=[1:3; "a"], b=[11:13; missing])
+
+            df = DataFrame(a=1:3, b=11:13)
+            with_logger(SimpleLogger(IOBuffer())) do
+                @test_throws MethodError pushfirst!(df, v, cols=cols, promote=false)
+            end
+            @test pushfirst!(df, v, cols=cols) ≅ DataFrame(a=["a"; 1:3], b=[missing; 11:13])
+
+            df = DataFrame(a=1:3, b=11:13)
+            with_logger(SimpleLogger(IOBuffer())) do
+                @test_throws MethodError insert!(df, 2, v, cols=cols, promote=false)
+            end
+            @test insert!(df, 2, v, cols=cols) ≅ DataFrame(a=[1, "a", 2, 3], b=[11, missing, 12, 13])
+        end
+    end
+end
+
+@testset "push!/pushfirst!/insert! with :setequal and wrong number of entries" begin
+    df = DataFrame(a=1:3)
+    @test_throws ArgumentError push!(df, (a=10, b=20))
+    @test_throws ArgumentError push!(df, "a")
+    @test_throws ArgumentError pushfirst!(df, (a=10, b=20))
+    @test_throws ArgumentError pushfirst!(df, "a")
+    @test_throws ArgumentError insert!(df, 2, (a=10, b=20))
+    @test_throws ArgumentError insert!(df, 2, "a")
+end
+
+@testset "push!/pushfirst!/insert! with self" begin
+    df = DataFrame(a=1:3, b=2:4, c=3:5)
+    @test push!(df, df[2, :]) == DataFrame(a=[1:3; 2], b=[2:4; 3], c=[3:5; 4])
+    @test pushfirst!(df, df[3, :]) == DataFrame(a=[3; 1:3; 2], b=[4; 2:4; 3], c=[5; 3:5; 4])
+    @test insert!(df, 3, df[1, :]) == DataFrame(a=[3; 1; 3; 2:3; 2], b=[4; 2; 4; 3:4; 3], c=[5; 3; 5; 4:5; 4])
+    df = DataFrame(a=1:3, b=2:4)
+    df.c = df.a
+    @test_throws AssertionError push!(df, df[2, :])
+    @test_throws AssertionError pushfirst!(df, df[2, :])
+    @test_throws AssertionError insert!(df, 2, df[2, :])
+    @test df == DataFrame(a=1:3, b=2:4, c=1:3)
+end
+
+@testset "multicolumn aliasing" begin
```

Comment:
@nalimilan - this test is made to additionally make sure that we correctly recover after error in complex multi-alias cases.